packages/core/src/storage/extract.ts

```diff
@@ -17,62 +22,98 @@ export function extractStorageLayout(
   contractDef: ContractDefinition,
   decodeSrc: SrcDecoder,
   deref: ASTDereferencer,
+  storageLayout?: StorageLayout | undefined,
 ): StorageLayout {
-  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion };
+  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion, flat: false };
+  if (storageLayout !== undefined) {
+    layout.types = storageLayout.types;
+    if (layout.types) {
+      for (const type of Object.entries(layout.types)) {
+        if (type[1].members) {
+          for (const member of type[1].members) {
+            // If member includes contract and ast the declaration contract should be used, right now it always uses the contract being processed
+            if (typeof member === 'object' && 'contract' in member && member.contract) {
+              const [, contract] = getOriginContract(contractDef, member.astId, deref);
+              member.contract = contract;
+            }
+          }
+        } else {
+          // In some cases the past version didnt have members and the updated version has members = undefined,
+          // this is for those cases
+          type[1].members = undefined;
+        }
+      }
+    }
```

Comment:
I want this to use `mapValues` so we make a copy of the object instead of modifying it.

I think it can be written like this:

```typescript
    layout.types = mapValues(storageLayout.types, t => {
      let { label, members } = t;
      if (members) {
        members = members.map(m => pick(m, ['label', 'type']));
      }
      return { label, members };
    });
```

Reply:
Could not use mapValues on typescript so I did iterated over `Object.keys(storageLayout.types)` instead